File: roadtrip.py
# Solution to RoadTrip (HSPT 2021 - On Site)
# Author: Atharva Nagarkar

import heapq

# constant for infinity (large value)
oo = 1000000000

# read in test cases
tests = int(input())

for tc in range(1, tests + 1):

    # read in bounds for the problem
    n, m = map(int, input().split())

    # dict to store the node id for the string names
    nameToId = {}

    # the list of nodes for each alphabet 'A' -> 'J' + an extra "sink"
    nameList = [ [] for i in range(11) ]

    # list to store the graph as an adjacency list
    graph = [ [] for i in range(n + 1) ]

    # tables to store the best answer to get to this node
    dp = [ oo for i in range(n + 1) ]
    dist = [ 0 for i in range(n + 1) ]

    # read in the names and assign them ids
    for i in range(n):
        name = input().strip()
        nameToId[name] = i
        c = ord(name[0]) - ord('A')
        
        # add the node to its corresponding first-letter list
        nameList[c].append(i)

        # add each node to the "sink"
        graph[n].append((i, 0))

    # the only node for the "sink" letter is the "sink" node
    nameList[10].append(n)
    
    # read in the edges of the graph
    for i in range(m):
        parts = input().strip().split()

        # get the ids of the given locations
        u = nameToId[parts[0]]
        v = nameToId[parts[1]]

        # read in the weight of the edge
        w = int(parts[2])

        # add it as a bi-directional edge in our graph
        graph[u].append((v, w))
        graph[v].append((u, w))
    
    # function for a modified dijkstra's algorithm
    # this will tell us the shortest distance to reach each s-letter nodes from the t-letter nodes
    def dijk(s, t):

        # initially set the distances of each node to be infinity (not reached yet)
        for i in range(n + 1): dist[i] = oo
        
        # make our priority queue (which is implemented as a heapq)
        q = []
        # push the relevant nodes (t-letter nodes) to our priority queue
        for node in nameList[t]:
            # the weight of the node will be the minimum cost needed to reach here from the sink
            heapq.heappush(q, (dp[node], node))
            dist[node] = dp[node]
        
        # run dijkstra's algorithm
        while len(q) > 0:
            cost, node = heapq.heappop(q)
            if cost > dist[node]: continue
            for v, w in graph[node]:
                if dist[v] > cost + w:
                    dist[v] = cost + w
                    heapq.heappush(q, (dist[v], v))

        # update the answer for the s-letter nodes
        for node in nameList[s]:
            dp[node] = dist[node]
    
    # initially set the cost of reaching the "sink" node to be 0
    dp[n] = 0

    # now for each character from 'J' down to 'A', we have to run the
    # modified dijkstra's algorithm to find the best cost to go in between
    # adjacent letters in the alphabet
    for c in range(9, -1, -1):
        dijk(c, c + 1)
    
    # our answer will be the minimum distance to reach some "A"-letter node
    # from the sink
    ans = oo
    for node in nameList[0]:
        ans = min(ans, dp[node])
    
    # print it!
    print("{ans}".format(ans = ans))

File: roadtrip.java
import java.util.*;

public class roadtrip {
	// the number of letters we consider
	static final int numLets = 'J' - 'A' + 1;
	// a large number
	static final long big = Long.MAX_VALUE / 4;
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		// test cases
		int t = sc.nextInt();
		for(int tt = 1; tt <= t; ++tt) {
			// cities and roads
			int n = sc.nextInt(), m = sc.nextInt();
			// city names
			
			// we will create a map of each city name to its
			// respective index, to keep things simple with
			// numbers rather than strings
			String[] cities = new String[n];
			HashMap<String, Integer> cityIndex = new HashMap<>();
			for(int i = 0; i < n; ++i) {
				cities[i] = sc.next();
				cityIndex.put(cities[i], i);
			}
			
			// create our adjacency list
			ArrayDeque<edge>[] edges = new ArrayDeque[n];
			for(int i = 0; i < n; ++i) {
				edges[i] = new ArrayDeque<>();
			}
			
			for(int i = 0; i < m; ++i) {
				String cityA = sc.next(), cityB = sc.next();
				// use the map to get the respective indices
				int a = cityIndex.get(cityA), b = cityIndex.get(cityB);
				int w = sc.nextInt();
				
				// add edges in both directions
				edges[a].add(new edge(b, w));
				edges[b].add(new edge(a, w));
			}
			
			// the key to this solution is that we will split up each node
			// into 10 nodes, one for each letter between 'A' and 'J';
			// then, we will compute the shortest distance to each node for
			// each letter we have acquired along our shortest path;
			// this method is called a node-splitting Dijkstra's
			
			long[][] dist = new long[numLets][n];
			// initially we will fill our distance array with very large values
			for(long[] a : dist) {
				Arrays.fill(a, big);
			}
			// create a PriorityQueue of states
			PriorityQueue<state> pq = new PriorityQueue<>();
			
			// go through each city and consider candidates for our starting city
			// (those with the first letter A)
			for(int i = 0; i < n; ++i) {
				if(cities[i].charAt(0) == 'A') {
					// set the distance to zero, as we possibly start here
					dist[0][i] = 0;
					// add this as a start to the queue
					pq.add(new state(0, i, 0));
				}
			}
			
			// run our node-splitting dijkstra's
			while(!pq.isEmpty()) {
				state curr = pq.poll();
				
				// if we have already reached this letter and index with a better
				// distance, then there is no need to explore it again
				if(curr.d > dist[curr.let][curr.idx]) {
					continue;
				}
				
				// consider each outgoing edge
				for(edge e : edges[curr.idx]) {
					// compute new letter
					int newLet = curr.let;
					// if we are going to a city that has the next letter we need,
					// then increment our new letter
					if(cities[e.v].charAt(0) - 'A' == curr.let + 1) {
						++newLet;
					}
					
					// compute new distance
					long newD = curr.d + e.w;
					
					// see if this new distance is an improvement
					if(newD < dist[newLet][e.v]) {
						dist[newLet][e.v] = newD;
						pq.add(new state(newLet, e.v, newD));
					}
				}
			}
			
			// compute our answer
			long out = big;
			// for each node, see if reaching this node as the last node
			// in our road trip is the best answer
			for(int i = 0; i < n; ++i) {
				if(dist[numLets - 1][i] < out) {
					out = dist[numLets - 1][i];
				}
			}
			
			// print
			System.out.println(out);
		}
	}
	static class state implements Comparable<state>{
		int let, idx;
		long d;
		state(int ll, int ii, long dd){
			let = ll;
			idx = ii;
			d = dd;
		}
		public int compareTo(state in) {
			// sort by smaller distances first
			return Long.compare(d, in.d);
		}
	}
	static class edge{
		int v, w;
		edge(int vv, int ww){
			v = vv;
			w = ww;
		}
	}
}
/*

samples for convenience:

2
11 10
Alabama
Alaska
Buffalo
Columbia
Delaware
Elfville
Florida
Georgia
Hawaii
Idaho
Jupiter
Alabama Jupiter 1
Jupiter Alaska 2
Idaho Jupiter 1
Georgia Idaho 1
Hawaii Georgia 1
Florida Hawaii 1
Elfville Florida 1
Elfville Delaware 1
Delaware Columbia 1
Buffalo Columbia 1
10 9
Avalon
Harrison
Dover
Bermuda
Gillian
Camelot
Jackson
Elliot
Florida
Iguana
Dover Avalon 3
Avalon Gillian 10
Florida Avalon 71
Avalon Bermuda 5
Jackson Avalon 5
Avalon Harrison 52
Elliot Avalon 6
Iguana Avalon 4
Camelot Avalon 5

*/
